### PR TITLE
Edge autonomy enhancement: Endpoints can be updated when disconnect with apiserver

### PR DESCRIPTION
**Pull request by chenkaiyue**
**What type of PR is this?**
enhancement

**What this PR does**:
Endpoints can be updated when disconnect with apiserver.

**Target**
When the node is disconnected from the cloud and cannot obtain the latest service status from the apiserver, it can still filter and update the service back-end endpoints list according to the status of other nodes, which enhances the edge autonomous service update maintenance ability, ensures the availability of the service back-end, and avoids the failure of service access caused by the failure of other nodes.

Even in a more extreme case, when all nodes are disconnected from the cloud, they can still guarantee the availability of the service back end, greatly enhancing the autonomy of the cluster.

**Design**
When a node is disconnected from the cloud, if some other node fails at some time after disconnection, the pod corresponding to one or more endpoints running on the node cannot be accessed; however, because the lite-server on the node still caches the endpoint list of the service before disconnection, the access to the service will still be forwarded to the node which is abnormal, causing the failure of access;

At present, wrapper has the function of filtering endpoint by hijacking Kube-proxy. Edge-health maintains the state of nodes as Apiserver does. Besides Apiserver, edge-health can be used as another reference data source for node state acquisition.

image

As shown in the figure above, the node where the pod B is located fails. At the same time, in the worst case, all nodes are disconnected from the cloud side. The rest nodes can't get the information of the pod B failure through apiserver. The first node in the figure can get the information of the node failure through edge-health, and remove the pod B from the back end of the service to ensure the availability of service calls.

**Specific implementation**
1. Add an interface to provide node status in edge-health, and the wrapper will periodically access the interface to obtain node status and traverse the endpoints map, which will move endpoint belong to the node with abnormal status from subnet.Addresses to subnet.NotReadyAddresses and send modified watch event to endpoints channel.

2. For the service, node, endpoint and other data obtained from apiserver, filter these data according to the cached health status of nodes obtained from edge-health, and compare with the previous data, if there is any change, send the corresponding watch event to Kube-proxy.

**Some special scenarios**
1. The fault node which edge-health on other node thinks is actually connected with apiserver.
For service calls within the cluster, if the edge-health detection and access between nodes is abnormal, the access to the pod of this node is also likely to be blocked, so it can be filtered according to the state recognized by edge-health.

2. Node A is not disconnected from the cloud. There will be two sources of information at this time: apiserver and edge health. Meanwhile, node B can not access by other nodes, but it can still be connected to the cloud

Kube-proxy of node A will resync regularly. The apiserver on Node A thinks that endpoint on node B is normal, but the information obtained from edge-health thinks B is abnormal, and both information is periodic update, so there will be a jump;

In order to solve this situation, we believe that the edge-health result is more reliable (as described in 1). Therefore, it is not only necessary to obtain the node state of edge-health regularly and update the endpoint state, but also filter the service, node, endpoint and other resources from apiserver according to the local cached node state, so as to ensure that the node status obtained by edge health is the standard.

3. The pod on abnormal node will be migrated to other nodes. The endpoint list will be updated and new endpoint will be added.
If a node is disconnected from cloud, only endpoints belong to abnormal nodes can be deleted through edge-health, new endpoints will not be added, but this does not affect the call of service.

**Components involved**
1. Edge-health node status query interface
2. Wrapper adds call of the edge-health and filter endpoints according to the node status of the edge-health (1. Filter from the list watch of apiserver; 2. Filter after call of edge-health)

**Test**

Deploy nginx deployment and corresponding service in a three node cluster to ensure that there is at least one pod on each node

***All nodes disconnected from cloud***
Disconnect all three nodes from the cloud, and disconnect one node from the other two nodes.
The current pod status:
```
[root@VM-0-124-centos ~]# kubectl get pod -o wide
NAME                        READY   STATUS    RESTARTS   AGE     IP              NODE           NOMINATED NODE   READINESS GATES
my-nginx-75897978cd-7x5t8   1/1     Running   0          18m     192.168.0.68    ecm-1xkcatq8   <none>           <none>
my-nginx-75897978cd-88ffn   1/1     Running   0          3h14m   192.168.0.3     ecm-hy1za18o   <none>           <none>
my-nginx-75897978cd-ckbn9   1/1     Running   0          3h6m    192.168.0.135   ecm-oz33tkf6   <none>           <none>
```
From the detected node state obtained by edge-health, we can see that the state of ecm-1xkcatq8 is already false
```
[root@VM-0-124-centos ~]# curl 127.0.0.1:51005/localinfo
{"ecm-1xkcatq8":{"Normal":false,"Hamc":"","Time":"0001-01-01T00:00:00Z"},"ecm-hy1za18o":{"Normal":true,"Hamc":"","Time":"0001-01-01T00:00:00Z"},"ecm-oz33tkf6":{"Normal":true,"Hamc":"","Time":"0001-01-01T00:00:00Z"}}
```
By looking at the routing table items on the current node, you can see that the original 3 back-end addresses have become 2, which is in line with the expectations
```
-A KUBE-SEP-22YW3O62YRV7OSGJ -s 192.168.0.3/32 -j KUBE-MARK-MASQ
-A KUBE-SEP-22YW3O62YRV7OSGJ -p tcp -m tcp -j DNAT --to-destination 192.168.0.3:80
-A KUBE-SEP-3ULPMXAOMPD3TGI5 -s 192.168.0.135/32 -j KUBE-MARK-MASQ
-A KUBE-SEP-3ULPMXAOMPD3TGI5 -p tcp -m tcp -j DNAT --to-destination 192.168.0.135:80
 
-A KUBE-SERVICES -d 172.16.50.158/32 -p tcp -m comment --comment "default/my-nginx: cluster IP" -m tcp --dport 80 -j KUBE-SVC-BEPXDJBUHFCSYIC3
-A KUBE-SERVICES -m comment --comment "kubernetes service nodeports; NOTE: this must be the last rule in this chain" -m addrtype --dst-type LOCAL -j KUBE-NODEPORTS
-A KUBE-SVC-BEPXDJBUHFCSYIC3 -m statistic --mode random --probability 0.50000000000 -j KUBE-SEP-3ULPMXAOMPD3TGI5
-A KUBE-SVC-BEPXDJBUHFCSYIC3 -j KUBE-SEP-22YW3O62YRV7OSGJ
```

***A,B Node is disconnected from the cloud, and node C is disconnected from nodes A and B***
This is the  special scenarios 1 above, we can see that the routing table on node a is also in line with the expectation
image

***Node A can connect with the cloud, and fault node B can connect with the cloud even though other nodes cannot access it***
The main verification is that our filtering logic filters the data obtained from apiserver except edge-health.Because node A is not disconnected from the cloud at this time, there will be two information sources, one is apiserver and the other is edge-health.
The situation seen from node A is also in line with expectations.
image

### CODE REVIEW

File: cmd/application-grid-wrapper/app/options/options.go [hunk 1]
```diff
@@ -29,6 +29,9 @@ type Options struct {
 	InsecureMode      bool
 	HostName          string
 	WrapperInCluster  bool
+	GetLocalNodeInfo  bool
+	HealthServerPort  int
+	HealthInternal    int
```
Comment (Beckham007) [inline cmd/application-grid-wrapper/app/options/options.go:34]: Internal -> Interval or Period?

Comment (chenkaiyue) [reply cmd/application-grid-wrapper/app/options/options.go:34]: yes, is interval...

File: pkg/application-grid-wrapper/storage/cache.go [hunk 2]
```diff
@@ -135,11 +141,38 @@ func (sc *storageCache) GetNode(hostName string) *v1.Node {
...
+
 // rebuildEndpointsMap updates all endpoints stored in storageCache.endpointsMap dynamically and constructs relevant modified events
 func (sc *storageCache) rebuildEndpointsMap() []watch.Event {
 	evts := make([]watch.Event, 0)
 	for name, endpointsContainer := range sc.endpointsMap {
-		newEps := pruneEndpoints(sc.hostName, sc.nodesMap, sc.servicesMap, endpointsContainer.endpoints, sc.wrapperInCluster)
+		newEps := pruneEndpoints(sc.hostName, sc.nodesMap, sc.servicesMap, endpointsContainer.endpoints, sc.localNodeInfo, sc.wrapperInCluster, sc.serviceAutonomyEnhancementSwitch)
 		if apiequality.Semantic.DeepEqual(newEps, endpointsContainer.modified) {
 			continue
 		}
```
Comment (neweyes) [inline pkg/application-grid-wrapper/storage/cache.go:178]: please add lock for operation ”sc.endpointsMap[name].modified = newEps“

Comment (chenkaiyue) [reply pkg/application-grid-wrapper/storage/cache.go:178]: Ok, i will create another PR to fix this.

File: cmd/application-grid-wrapper/app/options/service-autonomy-enhancement-options.go [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+		}
+		arr := strings.Split(para, "=")
+		trimkey := strings.TrimSpace(arr[0])
+		switch trimkey {
+		case "address":
+			(*p).NeighborStatusSvc = strings.TrimSpace(arr[1])
+		case "interval":
+			interval, _ := strconv.Atoi(strings.TrimSpace(arr[1]))
+			(*p).UpdateInterval = interval
+		case "switch":
```
Comment (neweyes) [inline cmd/application-grid-wrapper/app/options/service-autonomy-enhancement-options.go:34]: change this ”switch“ to "enabled"?

Comment (chenkaiyue) [reply cmd/application-grid-wrapper/app/options/service-autonomy-enhancement-options.go:34]: OK~

**Review by neweyes — APPROVED** [review]

**Pull request merged**